### PR TITLE
Add a feature gate for allowing integers as the address of references in constants

### PR DESCRIPTION
**Pull request by oli-obk**
implementation for #63197

cc @rust-lang/wg-const-eval

### CODE REVIEW

File: src/librustc_mir/interpret/validity.rs [hunk 1]
```diff
@@ -382,14 +382,20 @@ impl<'rt, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> ValidityVisitor<'rt, 'mir, '
...
             // alignment should take attributes into account).
             .unwrap_or_else(|| (place.layout.size, place.layout.align.abi));
+        let ptr = self.ecx.memory.check_ptr_access_align(
+            place.ptr,
+            size,
+            Some(align),
+            CheckInAllocMsg::InboundsTest,
+        );
+        if let Err(InterpErrorInfo { kind: err_unsup!(ReadBytesAsPointer), ..}) = ptr {
+            if self.ecx.tcx.features().const_int_ref {
```
Comment (RalfJung) [inline src/librustc_mir/interpret/validity.rs:392]: Wow what a hack. This needs *way* more comments (more than "no explanation at all", anyway. ;)

But I guess the goal here actually *is* to allow pointers that CTFE would consider dangling, so no surprise I guess.

File: src/librustc_mir/interpret/validity.rs [hunk 2]
```diff
@@ -382,14 +382,20 @@ impl<'rt, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> ValidityVisitor<'rt, 'mir, '
...
+                return Ok(());
+            }
+        }
         // Direct call to `check_ptr_access_align` checks alignment even on CTFE machines.
```
Comment (RalfJung) [inline src/librustc_mir/interpret/validity.rs:396]: Please move this comment above the `check_ptr_access_align` call again.

File: src/librustc_mir/interpret/validity.rs (hunk 1 above)
Comment (RalfJung) [reply src/librustc_mir/interpret/validity.rs:392]: Oh also, feature gates shouldn't affect what we do in Miri, so please make this conditional CTFE-only.

File: src/librustc_mir/interpret/validity.rs [hunk 3]
```diff
@@ -405,7 +411,7 @@ impl<'rt, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> ValidityVisitor<'rt, 'mir, '
             err_ub!(PointerOutOfBounds { .. }) =>
                 { "a dangling {} (going beyond the bounds of its allocation)", kind },
             err_unsup!(ReadBytesAsPointer) =>
-                { "a dangling {} (created from integer)", kind },
+                { "a dangling {} (created from integer). Add `#![feature(const_int_ref)]` to the crate attributes to enable", kind },
```
Comment (RalfJung) [inline src/librustc_mir/interpret/validity.rs:414]: We shouldn't show this feature gate help on stable...

File: src/test/ui/consts/const-int-ref.stderr [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+LL | |     int: usize,
+LL | |     ptr: &'static i32
+LL | | }
+LL | |
+LL | | const GPIO: &i32 = unsafe { Transmute { int: 0x800 }.ptr };
+   | |___________________________________________________________^ consider adding a `main` function to `$DIR/const-int-ref.rs`
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0601`.
```
Comment (RalfJung) [inline src/test/ui/consts/const-int-ref.stderr:15]: This file shouldn't exist, right? It doesn't belong to any revision.

File: src/test/ui/consts/const-int-ref.rs [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+#[repr(C)]
+union Transmute {
+    int: usize,
+    ptr: &'static i32
+}
+
+const GPIO: &i32 = unsafe { Transmute { int: 0x800 }.ptr };
+//[vanilla]~^ ERROR it is undefined behavior
+
+fn main() {}
```
Comment (RalfJung) [inline src/test/ui/consts/const-int-ref.rs:17]: Please add a test making sure we still catch NULL and unaligned references.

File: src/librustc_feature/active.rs [hunk 6]
```diff
@@ -565,6 +565,9 @@ declare_features! (
     /// Allow conditional compilation depending on rust version
     (active, cfg_version, "1.45.0", Some(64796), None),
 
+    /// Allows constants of reference type to have integers for the address.
+    (active, const_int_ref, "1.45.0", Some(63197), None),
```
Comment (RalfJung) [inline src/librustc_feature/active.rs:569]: Shouldn't there be a dedicated tracking issue? Not sure what the usual process is.

**Comment by RalfJung** [comment]
[This](https://github.com/rust-lang/rust/issues/63197#issuecomment-546053235) doesn't sound like there is lang team consensus for this so far. But maybe nightly-only experimentation is okay?

File: src/test/ui/consts/const-int-ref.stderr (hunk 4 above)
Comment (oli-obk) [reply src/test/ui/consts/const-int-ref.stderr:15]: hmm... I thought tidy warns us about such files. Yea this is leftover from a previous non-rev version

File: src/librustc_feature/active.rs (hunk 6 above)
Comment (oli-obk) [reply src/librustc_feature/active.rs:569]: We can change that issue to a tracking issue

File: src/test/ui/consts/const-int-ref.stderr (hunk 4 above)
Comment (RalfJung) [reply src/test/ui/consts/const-int-ref.stderr:15]: I think tidy doesn't understand revisions, it warns only about entirely stale `stderr` files.

**Comment by oli-obk** [comment]
Hmm yea, I read that comment differently (finding it was what prompted me to write this PR). I think the best way forward with either interpretation is to have a nightly feature for it so we can show what's going on. Especially considering the small change necessary, making it trivial to remove the unstable feature again, I think we should definitely create a feature gate and then prompt T-lang again for discussion of the feature. I can create a writeup with examples that run on the playground once this PR is merged.

**Comment by RalfJung** [comment]
> I think the best way forward with either interpretation is to have a nightly feature for it so we can show what's going on. Especially considering the small change necessary, making it trivial to remove the unstable feature again, I think we should definitely create a feature gate and then prompt T-lang again for discussion of the feature. I can create a writeup with examples that run on the playground once this PR is merged.

Makes sense.

> We can change that issue to a tracking issue

If you do, we'll need to start collecting concerns in the main issue so we have them centralized. Not sure if editing the OP or creating a new issue is better.

My main concern is:

The "obvious" alternative is to just use raw pointers, so the main question is -- is not having that `unsafe` block on every ptr use worth allowing potentially dangling references through our const-eval checks? (I keep changing my mind on that, right now I feel it might not be.)

This is briefly mentioned [here](https://github.com/rust-lang/rust/issues/63197#issuecomment-517420231), and the response makes me think this is a raw ptr ergonomics issue and maybe relaxing reference checks isn't the best solution.

**Comment by RalfJung** [comment]
Also note https://github.com/rust-lang/rust/issues/63359 which is mentioned somewhere in https://github.com/rust-lang/rust/issues/63197 -- could you extend your PR to handle that as well?

File: src/librustc_mir/interpret/validity.rs (hunk 1 above)
Comment (RalfJung) [reply src/librustc_mir/interpret/validity.rs:392]: > please make this conditional CTFE-only.

On this note, do we have tests in Miri that make sure we still catch dangling integer references (and fn pointers)?

File: src/librustc_feature/active.rs (hunk 6 above)
Comment (RalfJung) [reply src/librustc_feature/active.rs:569]: Reading that issue, it contains so much unrelated discussion, I'd prefer a new clean tracking issue.

File: src/librustc_mir/interpret/validity.rs [hunk 7]
```diff
@@ -410,8 +415,15 @@ impl<'rt, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> ValidityVisitor<'rt, 'mir, '
                 { "a dangling {} (address 0x{:x} is unallocated)", kind, i },
             err_ub!(PointerOutOfBounds { .. }) =>
                 { "a dangling {} (going beyond the bounds of its allocation)", kind },
-            err_unsup!(ReadBytesAsPointer) =>
-                { "a dangling {} (created from integer). Add `#![feature(const_int_ref)]` to the crate attributes to enable", kind },
+            err_unsup!(ReadBytesAsPointer) => {
+                "a dangling {} (created from integer).{}",
```
Comment (RalfJung) [inline src/librustc_mir/interpret/validity.rs:419]: The `.` should also be moved to the nightly-only part.

**Comment by oli-obk** [comment]
Closing after https://github.com/rust-lang/rust/issues/63197#issuecomment-634538553 has convinced me that this is the wrong thing to do and we should instead suggest raw pointers once raw borrows are stable and thus a viable alternative.

**Pull request closed without merging**